src/NumericalAlgorithms/Interpolation/LinearInterpolationND.hpp

```diff
@@ -0,0 +1,374 @@
+            typename CompressedIndex_t, typename X1, typename... Xt>
+  struct RecursiveInterpolationKernel {
+    static inline double const interpolate(
+        Data_t const& data, std::array<size_t, Dimension> const& indexArray,
+        Xt const&... xin, X1 const& x1) {
+      auto const xA = data.x[whichDimension][indexArray[whichDimension]];
+      auto const xB = data.x[whichDimension][indexArray[whichDimension] + 1];
+
+      auto const A = RecursiveInterpolationKernel<
+          Data_t, whichDimension - 1, whichVariable,
+          CompressedIndexRecursion_t<CompressedIndex_t, 0>,
+          Xt...>::interpolate(data, indexArray, xin...);
+      auto const B = RecursiveInterpolationKernel<
+          Data_t, whichDimension - 1, whichVariable,
+          CompressedIndexRecursion_t<CompressedIndex_t, 1>,
+          Xt...>::interpolate(data, indexArray, xin...);
+
+      if constexpr (UniformSpacing) {
+        return linearInterpolation1D_Uniform(
+            A, B, xA, xB, data.inverseSpacing[whichDimension], x1);
+      } else {
+        return linearInterpolation1D_General(A, B, xA, xB, x1);
+      }
+    }
+  };
+
+  // Unpack in dimension!
+  template <typename Data_t, size_t whichVariable, typename CompressedIndex_t,
+            typename X1>
+  struct RecursiveInterpolationKernel<Data_t, 0, whichVariable,
+                                      CompressedIndex_t, X1> {
+    static inline double const interpolate(
+        Data_t const& data, std::array<size_t, Dimension> const& indexArray,
+        X1 const& x1) {
+      size_t constexpr whichDimension = 0;
+
+      auto const xA = data.x[whichDimension][indexArray[whichDimension]];
+      auto const xB = data.x[whichDimension][indexArray[whichDimension] + 1];
+
+      auto const A =
+          data.y[whichVariable +
+                 NumberOfVariables *
+                     CompressedIndexRecursion_t<CompressedIndex_t, 0>::value(
+                         indexArray, data.numberOfPoints)];
+      auto const B =
+          data.y[whichVariable +
+                 NumberOfVariables *
+                     CompressedIndexRecursion_t<CompressedIndex_t, 1>::value(
+                         indexArray, data.numberOfPoints)];
+
+      if constexpr (UniformSpacing) {
+        return linearInterpolation1D_Uniform(
+            A, B, xA, xB, data.inverseSpacing[whichDimension], x1);
+      } else {
+        return linearInterpolation1D_General(A, B, xA, xB, x1);
+      }
+    }
+  };
+
+  // Unpack in dimension!
+  template <typename Data_t, size_t whichDimension, typename CompressedIndex_t,
+            typename X1, typename... Xt>
+  struct RecursiveInterpolationKernelGeneral {
+    static inline double const interpolate(
+        Data_t const& data, std::array<size_t, Dimension> const& indexArray,
+        size_t which_var, Xt const&... xin, X1 const& x1) {
+      auto const xA = data.x[whichDimension][indexArray[whichDimension]];
```

Comment:
We've found that the recursion is (surprisingly) slow and instead opted for different definitions depending on the `Dimension` in other contexts (see, e.g. `collapsed_index` in src/DataStructures/Index.hpp). This is also what the LANL interpolation library does (I forget what it's called now)